pkg/localstore/sampler.go

```diff
@@ -0,0 +1,197 @@
+	TotalIterated     atomic.Int64
+	NotFound          atomic.Int64
+	IterationDuration atomic.Int64
+	GetDuration       atomic.Int64
+	HmacrDuration     atomic.Int64
+}
+
+func (s *sampleStat) String() string {
+	return fmt.Sprintf(
+		"Total: %d NotFound: %d Iteration Durations: %d secs GetDuration: %d secs HmacrDuration: %d",
+		s.TotalIterated.Load(),
+		s.NotFound.Load(),
+		s.IterationDuration.Load()/1000000,
+		s.GetDuration.Load()/1000000,
+		s.HmacrDuration.Load()/1000000,
+	)
+}
+
+// ReserveSample generates the sample of reserve storage of a node required for the
+// storage incentives agent to participate in the lottery round. In order to generate
+// this sample we need to iterate through all the chunks in the node's reserve and
+// calculate the transformed hashes of all the chunks using the anchor as the salt.
+// In order to generate the transformed hashes, we will use the std hmac keyed-hash
+// implementation by using the anchor as the key. Nodes need to calculate the sample
+// in the most optimal way and there are time restrictions. The lottery round is a
+// time based round, so nodes participating in the round need to perform this
+// calculation within the round limits.
+// In order to optimize this we use a simple pipeline pattern:
+// Iterate chunk addresses -> Get the chunk data and calculate transformed hash -> Assemble the sample
+func (db *DB) ReserveSample(ctx context.Context, anchor []byte, storageDepth uint8) (storage.Sample, error) {
+
+	g, ctx := errgroup.WithContext(ctx)
+	addrChan := make(chan swarm.Address)
+	var stat sampleStat
+	logger := db.logger.WithName("sampler").V(1).Register()
+
+	// Phase 1: Iterate chunk addresses
+	g.Go(func() error {
+		defer close(addrChan)
+		iterationStart := time.Now()
+		err := db.pullIndex.Iterate(func(item shed.Item) (bool, error) {
+			select {
+			case addrChan <- swarm.NewAddress(item.Address):
+				stat.TotalIterated.Inc()
+				return false, nil
+			case <-ctx.Done():
+				return true, ctx.Err()
+			case <-db.close:
+				return true, errDbClosed
+			}
+		}, &shed.IterateOptions{
+			StartFrom: &shed.Item{
+				Address: generateAddressAt(db.baseKey, int(storageDepth)),
+			},
+		})
+		if err != nil {
+			logger.Error(err, "sampler: failed iteration")
+			return err
+		}
+		stat.IterationDuration.Add(time.Since(iterationStart).Microseconds())
+		return nil
+	})
+
+	// Phase 2: Get the chunk data and calculate transformed hash
+	sampleItemChan := make(chan swarm.Address)
+	const workers = 6
+	for i := 0; i < workers; i++ {
+		g.Go(func() error {
+			hmacr := hmac.New(swarm.NewHasher, anchor)
+
+			for addr := range addrChan {
+				getStart := time.Now()
+				chItem, err := db.get(ctx, storage.ModeGetSync, addr)
+				stat.GetDuration.Add(time.Since(getStart).Microseconds())
+				if err != nil {
+					stat.NotFound.Inc()
+					continue
+				}
+
+				hmacrStart := time.Now()
+				_, err = hmacr.Write(chItem.Data)
+				if err != nil {
+					return err
+				}
```

Comment:
I think this may be like this but if we ever want to inclusion proof sample items it may be better to use Single Segment Transform with BMT instead of HMAC

Reply:
So I thought with the BMT hasher we should be able to verify 32byte segments (hmac results) no?